LogicalQ/Experiments.py

```diff
@@ -358,6 +373,144 @@ def save_progress():
+    
+    def save_progress():
+        pickle.dump(lg_all_data, lg_save_file, protocol=5)
+        lg_save_file.close()
+        pickle.dump(phys_all_data, phys_save_file, protocol=5)
+        phys_save_file.close()
+        
+    atexit.register(save_progress)
+
+    if with_mp:
+        # @TODO Perform rigorous performance test to ensure implementation is fast and robust. There is not enough RAM on my laptop to do multiprocessing, especially since this function bundles in scaling testing for LogicalCircuit objects. Perhaps when PACE access comes online
+        
+        exp_inputs_list = [
+            (
+                (n_qubits, circuit_length),
+                circuit_factory(n_qubits=n_qubits, circuit_length=circuit_length),
+                noise_model_factory(n_qubits=n_qubits),
+                backend, method, shots, stabilizer_code
+            )
+            for (n_qubits, circuit_length) in itertools.product(range(min_n_qubits, max_n_qubits+1), range(min_circuit_length, max_circuit_length+1))
+        ]
+
+        cpu_count = (os.cpu_count() - 2) or 1
+
+        # batch_size = max(int(np.ceil((max_n_qubits+1-min_n_qubits)*(max_circuit_length+1-min_circuit_length)/cpu_count)), 1)
+        print(f"Applying multiprocessing to {len(exp_inputs_list)} samples across {cpu_count} CPUs")
+
+        start = time.perf_counter()
+
+        with Pool(cpu_count) as pool:
+            #results_iterator = pool.imap_unordered(parity_experiment_core), exp_inputs_list)
+
+            #mp_result = list(tqdm(results_iterator, total=len(exp_inputs_list)))
+            
+            mp_result = pool.starmap(
+                parity_experiment_core,
+                exp_inputs_list #*[list(exp_inputs) for exp_inputs in zip(*exp_inputs_list)],
+                # chunksize=batch_size
+            )
+
+            # Unzip results
+            for (task_id, phys_result, lg_result) in mp_result:
+                phys_all_data[task_id[0]][task_id[1]] = phys_result#[0]
+                if task_id[0] <= 2:
+                    lg_all_data[task_id[0]][task_id[1]] = lg_result#[0]
+
+        stop = time.perf_counter()
+    else:
+        start = time.perf_counter()
+
+        prog = 0
+        prog_max = (max_n_qubits - min_n_qubits + 1) * (max_circuit_length - min_circuit_length + 1)
+
+        for n_qubits in range(min_n_qubits, max_n_qubits+1):
+            phys_sub_data = {}
+            lg_sub_data = {}
+
+            noise_model_n = noise_model_factory(n_qubits=n_qubits)
+
+            for circuit_length in range(min_circuit_length, max_circuit_length+1):
+                
+                prog += 1
+                
+                # @TODO make this less jank
+                print(f"Estimated Progress: {np.round(100 * prog / prog_max, 2)}%")
+                
+                circuit = circuit_factory(n_qubits=n_qubits, circuit_length=circuit_length)
+                
+                if n_qubits <= 2: # Don't go above 2 b/c of RAM requirements
```

Comment:
I don't think we should put these sorts of checks in at the experiment level, I feel like it just adds a limitation which is on the user's end on not the code end. The user should be allowed to determine how many qubits they can run themselves and design experiments based on that.